[Geo] Add Well Known Text (WKT) Parsing Support to ShapeBuilders

nknize opened the pull request:
This PR adds WKT support to Geo ShapeBuilders.

This supports the following format:

```
POINT (30 10)
LINESTRING (30 10, 10 30, 40 40)
POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))
POLYGON ((35 10, 45 45, 15 40, 10 20, 35 10), (20 30, 35 35, 30 20, 20 30))
MULTIPOINT ((10 40), (40 30), (20 20), (30 10))
MULTIPOINT (10 40, 40 30, 20 20, 30 10)
MULTILINESTRING ((10 10, 20 20, 10 40),(40 40, 30 30, 40 20, 30 10))
MULTIPOLYGON (((30 20, 45 40, 10 40, 30 20)), ((15 5, 40 10, 10 20, 5 10, 15 5)))
MULTIPOLYGON (((40 40, 20 45, 45 30, 40 40)), ((20 35, 10 30, 10 10, 30 5, 45 20, 20 35), (30 20, 20 15, 20 25, 30 20)))
```
closes #9120

nknize commented:
@elasticmachine please test this.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        throw new IllegalArgumentException("Unknown geometry type: " + type);
+    }
+
+    private static EnvelopeBuilder parseBBox(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
```
This will also accept something like "BBOX (-10) 10) -10) 10)", I think this should throw an error. Its a bit constructed but I think the parser shouldn't be that lenient.

nknize replied:
:+1:

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        }
+        PointBuilder pt = new PointBuilder(nextNumber(stream), nextNumber(stream));
+        if (isNumberNext(stream) == true) {
+            nextNumber(stream);
```
Is this a potential Z value? So we support parsing it but drop it?

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+            throws IOException, ElasticsearchParseException {
+        CoordinatesBuilder coordinates = new CoordinatesBuilder();
+        coordinates.coordinate(parseCoordinate(stream));
+        while (nextCloserOrComma(stream).equals(COMMA)) {
```
Close or Comma will also again allow for things like "( 1, 2, 3) 4 )" being accepted, no?

nknize replied:
Cleaned up coordinate list parsing to better handle coordinates with and without parens.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        if (nextEmptyOrOpen(stream).equals(EMPTY)) {
+            return null;
+        }
+        // todo orientation?
```
What needs to be done here?

nknize replied:
Actually, nothing. This is handled in the builder. Removing todo.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        }
+        // todo orientation?
+        PolygonBuilder builder = new PolygonBuilder(parseLine(stream), ShapeBuilder.Orientation.RIGHT);
+        if (nextWord(stream).equals(LPAREN)) {
```
What if next word is something else than LPAREN?

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        // todo orientation?
+        PolygonBuilder builder = new PolygonBuilder(parseLine(stream), ShapeBuilder.Orientation.RIGHT);
+        if (nextWord(stream).equals(LPAREN)) {
+            while (nextCloserOrComma(stream).equals(COMMA)) {
```
same problem as above I think, only that unmatched closing brackets will create holes here.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+            case StreamTokenizer.TT_EOL: return EOL;
+            case StreamTokenizer.TT_NUMBER: return NUMBER;
+        }
+        return "'" + (char)stream.ttype + "'";
```
nit: whitespace

nknize commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,155 @@
+        List<Coordinate> coordinates = new ArrayList<>(numPoints);
+        GeoPoint p;
+        for (int i=0; i<numPoints; ++i) {
+            p = RandomShapeGenerator.randomPointIn(random(), -90d, -90d, 90d, 90d);
```
This is used for testing `LineStringBuilder` and `MultiLineStringBuilder`. If I allow points to cross the dateline, a `MultiLineStringBuilder` is always created. To keep testing simple, I limited the random generation of linestrings to not cross the dateline. This gave me control over testing `LineStrings` and `MultiLineStrings` separately.

nknize commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Override
+    public void testParseLineString() throws IOException {
```
Simple fix... the random point generation was set to `randomIntBetween(0, 100)`. 0 will throw an empty/invalid points exception.

nknize commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+        throw new ElasticsearchParseException("expected " + RPAREN + " but found: " + tokenString(stream), stream.lineno());
+    }
+
+    private static String nextCloserOrComma(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
```
I actually meant closer. As in a closer parentheses. But will change if this is confusing.

cbuescher replied:
Ok, I see. Nevermind, since its a private method I leave it up to you to change, I was confused a bit but the method is short enough to understand what its doing.

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java:
```diff
@@ -339,6 +340,47 @@ public String toString() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(coordinate.x + GeoWKTParser.SPACE + coordinate.y);
+        if (Double.isNaN(coordinate.z) == false) {
+            sb.append(GeoWKTParser.SPACE + coordinate.z);
```
Not yet... will be handled in #25738

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java:
```diff
@@ -339,6 +340,47 @@ public String toString() {
+        return sb.toString();
+    }
+
+    private static String coordinateToWKT(final Coordinate coordinate) {
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java:
```diff
@@ -136,6 +137,23 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        if (shapes.isEmpty()) {
+            sb.append(GeoWKTParser.EMPTY);
+        } else {
+            if (shapes.isEmpty() == false) {
```
bonehead typo

nknize commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Override
+    public void testParseMultiPoint() throws IOException {
```
Fixed... the random point generation was set to randomIntBetween(0, 100). 0 will throw an empty/invalid points exception.

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -273,10 +278,16 @@ CoordinateNode validate(CoordinateNode coordinates, boolean coerce) {
             shapeTypeMap.put(type.shapename, type);
+            // add 'BBOX' wkt name support for Envelope
+            if (type.equals(ENVELOPE)) {
+                shapeTypeMap.put(ENVELOPE.wktName().toLowerCase(Locale.ROOT), type);
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -273,10 +278,16 @@ CoordinateNode validate(CoordinateNode coordinates, boolean coerce) {
+    private static final String BBOX;
 
     static {
+        BBOX = "BBOX";
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -300,6 +311,10 @@ public abstract ShapeBuilder getBuilder(CoordinateNode coordinates, DistanceUnit
                                             ShapeBuilder.Orientation orientation, boolean coerce);
     abstract CoordinateNode validate(CoordinateNode coordinates, boolean coerce);
 
+    public String wktName() {
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java:
```diff
@@ -339,6 +340,47 @@ public String toString() {
+
+    public String toWKT() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(type().wktName());
```
Per the WKT spec, casing doesn't matter

nknize commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+            tokenizer.wordChars('+', '+');
+            tokenizer.wordChars('.', '.');
+            tokenizer.whitespaceChars(0, ' ');
+            tokenizer.commentChar('#');
```
:+1:  Just push a commit to test for comments

nknize commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,279 @@
+            case ENVELOPE:
+                return parseBBox(stream);
+        }
+        throw new IllegalArgumentException("Unknown geometry type: " + type);
```
:+1:  done

nknize commented:
Thanks for the review @cbuescher 

I went ahead and added more strict parsing (to include better paren parsing for coordinate lists). I also added tests for comments and illegal geometry types. I think this is close. If you wouldn't mind giving another review. I will try to add geometry collection testing before committing.

Also, since this is a fresh feature, what we can do is set this as experimental for 6.1 and flush it out in bug fixes. Thoughts?

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,182 @@
+            RandomShapeGenerator.createShape(random(), RandomShapeGenerator.ShapeType.POLYGON));
+        Coordinate[] coords = builder.coordinates()[0][0];
+        LinearRing shell = GEOMETRY_FACTORY.createLinearRing(coords);
+        Polygon expected = GEOMETRY_FACTORY.createPolygon(shell, null);
```
Should this also work with holes? I just tried parsing the second polygon example from the issue decription and it fails (`POLYGON ((35 10, 45 45, 15 40, 10 20, 35 10), (20 30, 35 35, 30 20, 20 30))`)  I suspect the second argument is a hole so I was trying to see of this was tested, seems like this needs to be added still.

nknize replied:
:+1:  Good catch! I agree we should have tests for holes. This is a widely used use case. Will add that next.

nknize replied:
I added an explicit test for polygons with holes below in `testParsePolygonWithHole`, are you asking for a random polygon with random hole? That's a bit more tricky to generate a random polygon with a random inscribed polygon as a hole. Probably not really necessary as long as we've tested at least one polygon with a hole?

cbuescher replied:
Yes, I saw the explicit test and I was hoping for a randomized polygon/hole thing, but I just realized that also GeoJsonShapeParserTests doesn't do this fully randomized. So if it is difficult I'm also fine if there is only the explicit test.

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    private void assertExpected(Shape expected, ShapeBuilder builder) throws IOException {
+        boolean isMalformed = randomBoolean();
```
I would prefer at least one good and one malformed case per test run. Otherwise we only catch bigger errors on multiple repetions and not e.g. on PR tests.

nknize replied:
I'll add an explicit malformed test for common typos.

nknize replied:
:+1:

cbuescher replied:
thanks

nknize commented:
@cbuescher just added explicit tests for the following:
* polygon with hole (can add one for multi polygon w/ and w/o holes)
* a malformed poly caused by common typos (can expand this test to include other common typos if we'd like).
* self crossing polygon

Will update documentation but I think this is close to ready. Let me know what you think.

cbuescher commented:
@nknize thanks for updating the test, if the polygon + hole randomization is tricky, let's leave that out. I think this is ready once some documentation is added.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java:
```diff
@@ -549,6 +630,32 @@ public boolean advanceExact(int target) throws IOException {
+    /** wraps a {@link Rectangle} to include an optional field name for queries */
+    public static class GeoBBox extends Rectangle {
+        /** optional fieldname for queries */
+        private String fieldName;
```
Why do we need the fieldname as part of the parsed bbox structure. As far as I can see this gets only parsed when an inner `field` parameter is seen, the use of which seems both undocumented and it also doesn't get rendered when using the GeoBoundingBoxQueryBuilder#doXContent method. I have a hope that we can remove this maybe?

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java:
```diff
@@ -465,6 +478,74 @@ public static GeoPoint parseGeoPoint(String data, GeoPoint point) {
+                currentFieldName = parser.currentName();
+                token = parser.nextToken();
+                if (FIELD_FIELD.match(currentFieldName)) {
+                    fieldName = parser.text();
```
Should this inner `field` parameter really still supported? I cannot find it mentioned in the docs and it doesn't seem to get written when using one of our builders. As far as I understand the whole bbox json should look like this:
```
"some.field.foo" : {
                        ... various formats defining the box ...
}
```

This is probably legacy and should be removed (probably in another PR?)

cbuescher commented:
@nknize I just did another round of reviews on the last commit adding WKT parsing support to the GeoBoundingBoxQueryBuilder. Does this need to be part of this PR? I agree it should be added to complete the WKT support (don't know what needs to be done with the other query builders that can contain shapes), but since the existing PR is already big as is (and was close to merging when the documentation is added)  I'd probably do this in a separate PR.

nknize commented:
@cbuescher I agree. After I made the push I figured it would be best to do it in a separate PR. So I'll go ahead and do that. Going to commit the updated docs here shortly and consider this PR done (after the doc review).

nknize commented:
@cbuescher added some WKT documentation to the `geo-shape.asciidoc`. If you're good with this I'll commit to 6.2 and 7.0 and open a separate PR for adding WKT support to GeoBoundingBox queries. Then move back to Z value support 
/cc @clintongormley

cbuescher commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -323,9 +360,22 @@ POST /example/doc
+--------------------------------------------------
+POST /example/doc
+{
+    "location" : "POLYGON ((100.0 0.0, 101.0 0.0, 101.0 1.0, 100.0 1.0, 100.0 0.0),
```
`gradle :docs:check` fails here, maybe because of the line break. Not sure how to fix though.

clintongormley replied:
If the line break is the problem, it's ok to have long lines.  The code blocks on the docs will scroll

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,306 @@
+    /** parse geometry from the stream tokenizer */
+    private static ShapeBuilder parseGeometry(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        final GeoShapeType type = GeoShapeType.forName(nextWord(stream));
+        switch (type) {
```
I get a warning here that GEOMETRYCOLLECTION is missing, looking at the added documentation this should be supported I think. If it is missing here as a type, we also need a test for it to catch this.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java:
```diff
@@ -0,0 +1,306 @@
+            case ENVELOPE:
+                return parseBBox(stream);
+        }
+        throw new IllegalArgumentException("Unknown geometry type: " + type);
```
nit: might throw in a DEFAULT case instead.

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -0,0 +1,243 @@
+        assertValidException(builder, IllegalArgumentException.class);
+    }
+
+    // todo: add geocollection
```
I think we need that test.

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -239,5 +240,26 @@ public void testInvalidGeometryType() throws IOException {
+        List<Coordinate> multiPointCoords = randomLineStringCoords();
+        MultiPointBuilder multiPointBuilder = new MultiPointBuilder(multiPointCoords);
+        GeometryCollectionBuilder builder = new GeometryCollectionBuilder()
+            .shape(polygonBuilder).shape(lineStringBuilder).shape(multiPointBuilder);
```
Can this be randomized so the GeometryCollectionBuilder has between 0 and e.g. 2 or 3 shapes?

nknize commented:
Thanks for such a great review @cbuescher  Looks like I've fixed the doc failures (at least on my test run, we'll see what elasticmachine has to say) and I added some randomization to GeometryCollection test. Pending everything passes green, I think this is good to go. Let me know what you think.

nknize commented:
@elasticmachine please test this

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -242,40 +242,7 @@ public void testInvalidGeometryType() throws IOException {
-            // assert empty shape collection
-            assertEquals(shapeCollection(expected).isEmpty(), builder.build().isEmpty());
-        }
+        GeometryCollectionBuilder gcb = RandomShapeGenerator.createGeometryCollection(random());
```
Great to have this, but it doesn't seem to test the no-shapes case. Would be nice to add this as another case (maybe wrapped in  `if (rarely())`, we don't need to test that edge case very often.

synhershko commented:
Yay! thanks!
